server/battles/BattleActionProcessor.cpp

```diff
@@ -1039,26 +1039,16 @@ void BattleActionProcessor::makeAttack(const CBattleInfoCallback & battle, const
 			totalKills += bsa.killedAmount;
 		}
 
-		{
-			MetaString text;
-			attacker->addText(text, EMetaText::GENERAL_TXT, 376);
-			attacker->addNameReplacement(text);
-			text.replaceNumber(totalDamage);
-			blm.lines.push_back(text);
-		}
+		addGenericDamageLog(blm, attackerState, totalDamage);
 
 		addGenericKilledLog(blm, defender, totalKills, multipleTargets);
 	}
 
 	// drain life effect (as well as log entry) must be applied after the attack
 	if(drainedLife > 0)
 	{
-		MetaString text;
-		attackerState->addText(text, EMetaText::GENERAL_TXT, 361);
-		attackerState->addNameReplacement(text, false);
-		text.replaceNumber(drainedLife);
-		defender->addNameReplacement(text, true);
-		blm.lines.push_back(std::move(text));
+		addGenericDrainedLifeLog(blm, attackerState, defender, drainedLife);
+		addGenericResurrectedLog(blm, attackerState, defender, static_cast<int64_t>(drainedLife / attackerState->health.getFirstHPleft()));
```

Comment:
Oh, missed this one.
I think it is incorrect?
`getFirstHPleft` is remaining hp of *first* unit. Not max HP of this unit type.
And even if replaced with correct method this still won't be correct.

If Vampire Lord with 30/40 HP drains 20 HP, 1 unit will be resurrected and be left with 10 HP. Even though drained HP was less than 40 (total HP of Vampire Lord).

Hmm.. Perhaps 'drainedLife' should not be int but rather simple struct `{ healthPoint, unitsCount }`?

Reply:
You are right... I used wrong method. Will fix.
Yet I left the mechanics unchanged. They heal by the amount of damage dealt resurrecting if its value covers all missing HP of wounded unit in the stack... or is this wrong ?

> LIFE_DRAIN
Affected unit will heal itself, resurrecting any dead units, by amount of dealt damage (Vampire Lord)

https://github.com/vcmi/vcmi/blob/5b7e355953d00bb863240a6d328bdfe9f97bc92f/docs/modders/Bonus/Bonus_Types.md#life_drain

Reply:
Mechanics is correct.
However knowing heal amount is not enough to determine how many units have resurrected.

For example, If first unit in your stack had 35/40 HP before attack then life drain for 20 HP would heal 5 HP to last unit and would resurrect one unit with 15 HP
But if first unit had 10/40 HP then life drain for 20 HP would only heal this unit to 30/40 HP and won't resurrect units.

Reply:
Ahhh I know now what you mean and where is the problem. After heal the first unit HP value is useless and the same goes for damageDealt. I was thinking ealier about adding one property to CHealth and a function getSingleHealResurrected() but maybe just counting units before and after heal inside makeAttack is fine ? There is only one more place where getSingleHealResurrected() would be used:
[Heal.cpp#L119](https://github.com/vcmi/vcmi/blob/5b7e355953d00bb863240a6d328bdfe9f97bc92f/lib/spells/effects/Heal.cpp#L119)

Reply:
> counting units before and after heal inside makeAttack is fine ?

If defender has Fire Shield he may be able to kill some of newly resurrected units. So we need to track actual number of resurrected units similar to how we track actual number of healed HP.

Perhaps:
- make applyBattleEffects return void instead of int64_t (which is not exactly logical / expected anyway)
- add structure similar to `FireShieldInfo & fireShield`, e.g. `LifeDrainInfo` and pass it as parameter to applyBattleEffects 
- this structure should contain int64_t that describes amount of drained health points and int32_t that describes number of resurrected units

Reply:
i think in H3 fire shield is applied after resurrection, so it is not accounted for

Reply:
It's easy to verify: just create a test map with respective units in oh3 editor, then you can test it in both oh3 and vcmi